src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3339,14 +3339,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
 
             (!sym.isModule || shouldAddAsModule) && (inBlock || !context.isInPackageObject(sym, context.owner))
           }
-          for (sym <- scope)
+          for (sym <- scope) if (shouldAdd(sym)) {
             // OPT: shouldAdd is usually true. Call it here, rather than in the outer loop
-            for (tree <- context.unit.synthetics.get(sym) if shouldAdd(sym)) {
+            for (tree <- context.unit.synthetics.get(sym)) { 
```

Comment:
To obey the comment, `synthetics.get(sym) match { case Some(tree) if shouldAdd(sym) => case _ => }`